Fix wrong use of alias when registering new hook

eternoendless opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | Hooking a module to hook `actionAuthenticationBefore` doesn't work on a clean install.
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | no... i think
| Deprecations? | no
| Fixed ticket? | Fixes #18945
| How to test?  | Read below

When installing a module that tried to hook to `actionAuthenticationBefore`, it didn't work.

After some investigation, we found out that said module wasn't previously registered in `ps_hooks` on a clean install (which is not an issue in itself because PrestaShop automatically adds them as needed). However, even if it's not registered in `ps_hooks`, it does have an alias in `ps_hook_alias`: `actionBeforeAuthentication`.

Whenever a hook doesn't exist in `ps_hooks`, it is automatically added. But if the hook has an alias, we found that the _lowercased_ alias was being inserted instead. Why? No clue. I think it was a bug. Maybe?

As far as I can tell, the bug goes way back to 2011, probably beyond...

---
**Update on 2020-08-05:** With this change, the alias behavior is now as follows:

**Context:**
- `someNewHook` is a given hook
- `someHook` is an alias for `someNewHook`

**Scenario 1:** Nominal case – a module registers to `someNewHook`, the hook `someNewHook` is dispatched
- table ps_hook contains “someNewHook”
- table ps_hook_module refers to “someNewHook”
- when `someNewHook` is dispatched, the module’s `hooksomeNewHook()` method is called

**Scenario 2:** Old code dispatches a renamed hook – a module registers to `someNewHook`, the hook `someHook` is dispatched
- table ps_hook contains “someNewHook”
- table ps_hook_module refers to “someNewHook”
- when `someHook` is dispatched, the module’s `hooksomeNewHook()` method is called

**Scenario 3:** Modules get backwards compatibility with renamed hooks – a module registers to `someHook`, the hook `someNewHook` is dispatched
- table ps_hook contains “someHook”
- table ps_hook_module contains “someHook”
- when `someNewHook` is dispatched, the module’s `hooksomeHook()` method is called

**Scenario 4:** Old code calls old hook and old module still works – a module registers to `someHook`, the hook `someHook` is dispatched
- table ps_hook contains “someHook”
- table ps_hook_module contains “someHook”
- when `someHook` is dispatched, the module’s `hooksomeHook()` method is called

---
**Update on 2020-08-06:** This change allowed me to realize that the the hook `displayHeader` was used as an alias of `Header` instead of the other way around. This was making `Header` be treated as the canonical hook name instead of an alias, which broke ps_shoppingcart.


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/12578)
<!-- Reviewable:end -->

Commit:
Fix wrong typehint

Commit:
Add comments and small changes with no functional impact

Commit:
Deprecate public method but keep functionality

Commit:
Test canonical hook name first, then alias (instead of the other way around)

Commit:
When creating a hook that doesn't exist in DB, make sure we use the canonical name -- and not the alias!

kpodemski commented on classes/Hook.php:
```diff
@@ -315,34 +314,67 @@ private static function getHookAliasesFor($hookName)
-        $aliases = Hook::getHookAliasesFor($hookName);
-        $aliases[] = $hookName;
+        $aliases = array_merge(
+            [$hookName],
```
@eternoendless i'm still lost whether we can or not use shorthand arrays :D

eternoendless replied:
We can!

PierreRambaud replied:
And we love it!

kpodemski commented:
it might be due to the fact that in 1.4 and before we had `hookHeader` etc. in 1.5 this have changed to `hookDisplayHeader` etc.

overall quite nice catch as i've never noticed this problem before 🤔

eternoendless commented:
> it might be due to the fact that in 1.4 and before we had `hookHeader` etc. in 1.5 this have changed to `hookDisplayHeader` etc.

Yeah, that's what aliases are for: you ask for the alias, and the canonical hook is issued/inserted. But I think that inserting the alias instead of the canonical name was a bug, because it would never work!

mickaelandrieu reviewed (dismissed):
Looks good but... regarding the importance of the classes impacted, maybe can we create some tests to confirm it works?

matks commented:
Indeed I think a behat test (behat because database is involved) would be good here. @eternoendless do you want me to add the test ?

eternoendless commented:
@matks Are you sure that this needs a Behat test? I think it would make more sense to build an _integration_ test much like the ones for the translator (which use a test module)

matks commented:
> @matks Are you sure that this needs a Behat test? I think it would make more sense to build an _integration_ test much like the ones for the translator (which use a test module)

Indeed I was thinking about an integration test. Behat might not be the best choice indeed.

matks commented:
🆙

Progi1984 commented:
Hello @eternoendless , thanks for your contribution, would you mind to rebase your pull request? Thanks

psyray commented:
Hi @eternoendless 
ping ! :)

Commit:
Extract getAllHookIds and add the option not to alias hooks to their canonical id

Commit:
Refactor getHookModuleExecList

Commit:
Clarify parameter & define return type in getHookAliasesFor

Commit:
Throw exception when hook does not exist in getNameById

Commit:
Use strict types in private methods

Commit:
Make sure alias hook registration/unregistration is tied to the alias and not the canonical hook

Commit:
Make sure the alias hook function is called if the hook was registered to an alias

Before this change, when `someHook` was invoked, modules having registered to `someHookAlias`
wouldn't be notified, because Hook would call `hookSomeHook()` (the method matching the
canonical hook) instead of `hookSomeHookAlias()` (the method matching the aliased hook,
present in the module). This change ensures that modules hooking to an alias receive the
appropriate method call.

Commit:
CS fixes

eternoendless commented:
Rebased on top of 1.7.7.x.

I was able to verify this change using the awesome [QA module](https://github.com/PrestaShop/ps_qualityassurance).

1. Register the hook using the module so that a log entry is saved when invoked:

<img width="601" alt="Screenshot 2020-08-04 at 10 40 48" src="https://user-images.githubusercontent.com/1009343/89275941-452dbe00-d643-11ea-86b1-6a26838027ca.png">

2. Verify that the new hook `actionAuthenticationBefore` has been added and registered instead of the alias `actionBeforeAuthentication` (this was the original bug) ✅ 

<img width="404" alt="Screenshot 2020-08-04 at 11 19 37" src="https://user-images.githubusercontent.com/1009343/89276872-6fcc4680-d644-11ea-8de8-2b7227a06767.png">

3. Try to log in as a customer on FO to trigger the hook (it doesn't matter if the customer exists)

<img width="1142" alt="Screenshot 2020-08-04 at 11 21 14" src="https://user-images.githubusercontent.com/1009343/89277077-a86c2000-d644-11ea-824e-0cc4e40daab2.png">

4. Verify that the hook was invoked by checking the logs ✅ 

<img width="1560" alt="Screenshot 2020-08-04 at 11 22 42" src="https://user-images.githubusercontent.com/1009343/89277255-e49f8080-d644-11ea-87a2-58215499f88b.png">

5. Now let's unregister the former hook and repeat the test using the alias `actionBeforeAuthentication` instead:

<img width="547" alt="Screenshot 2020-08-04 at 11 25 58" src="https://user-images.githubusercontent.com/1009343/89395596-d4070d00-d70d-11ea-8976-d4db292c9ec9.png">

6. Verify that the new hook `actionBeforeAuthentication` has been added and registered ✅ 

<img width="395" alt="Screenshot 2020-08-04 at 18 39 52" src="https://user-images.githubusercontent.com/1009343/89395723-00228e00-d70e-11ea-969f-57dbcc4cb73e.png">

7. Log in as a customer on FO to trigger the hook, then verify that the hook was invoked by checking the logs ✅ 

<img width="1552" alt="Screenshot 2020-08-05 at 11 10 46" src="https://user-images.githubusercontent.com/1009343/89395839-1d575c80-d70e-11ea-93da-9e06960248d6.png">

Commit:
Rename getHookAliasDictionary into getCanonicalHookNames

Commit:
Make isHookCallableOn public & extract method getMethodName

Commit:
Improve documentation

Commit:
Avoid listing aliases as hookable points

Commit:
Optimize query

+ CS fix

Commit:
Optimize query and improve cache id

Commit:
getHookAliasesFor no longer returns canonical hook names when given an alias

Commit:
Fix wrong hook alias in database (canonical name should be "displayHeader" and not the other way around)

PierreRambaud commented on classes/Hook.php:
```diff
@@ -175,44 +181,35 @@ public static function getHooks($position = false, $only_display_hooks = false)
+        $hook_ids = self::getAllHookIds($withAliases, $refreshCache);
 
-        return isset($hook_ids[$hook_name]) ? $hook_ids[$hook_name] : false;
+        return isset($hook_ids[$hookName]) ? $hook_ids[$hookName] : false;
```
```suggestion
        return $hook_ids[$hookName]) ?? false;
```

PierreRambaud commented on classes/Hook.php:
```diff
@@ -222,6 +219,11 @@ public static function getNameById($hook_id)
 							WHERE `id_hook` = ' . (int) $hook_id);
+
+            if (false === $result) {
+                throw new PrestaShopObjectNotFoundException('The hook id #%s does not exist in database', $hook_id);
```
```suggestion
                throw new PrestaShopObjectNotFoundException('The hook id #%d does not exist in database', $hook_id);
```

PierreRambaud commented on classes/Hook.php:
```diff
@@ -231,51 +233,55 @@ public static function getNameById($hook_id)
+     */
+    public static function isAlias(string $hookName): bool
+    {
+        $aliases = self::getCanonicalHookNames();
```
```suggestion
        $aliases = static::getCanonicalHookNames();
```

PierreRambaud commented on classes/Hook.php:
```diff
@@ -288,158 +294,219 @@ private static function getHookAliasesList()
+        $hooksToCheck = (!$strict) ? static::getAllKnownNames($hookName) : [$hookName];
+
+        foreach ($hooksToCheck as $currentHookName) {
+            if (is_callable([$module, self::getMethodName($currentHookName)])) {
```
```suggestion
            if (is_callable([$module, static::getMethodName($currentHookName)])) {
```

PierreRambaud commented on classes/Hook.php:
```diff
@@ -288,158 +294,219 @@ private static function getHookAliasesList()
+        // Since is_callable() will always return true when __call() is available,
+        // if the module was expecting an aliased hook name to be invoked, but we send
+        // the canonical hook name instead, the hook will never be acknowledged by the module.
+        $methodName = self::getMethodName($hookName);
```
```suggestion
        $methodName = static::getMethodName($hookName);
```

PierreRambaud commented on classes/Hook.php:
```diff
@@ -288,158 +294,219 @@ private static function getHookAliasesList()
-        $alias_list = Hook::getHookAliasList();
-        if (isset($alias_list[strtolower($hook_name)])) {
-            return $alias_list[strtolower($hook_name)];
+        $hookNamesByAlias = static::getCanonicalHookNames();
+        if (isset($hookNamesByAlias[strtolower($hookName)])) {
+            // return the canonical name (?)
+            return $hookNamesByAlias[strtolower($hookName)];
```
```suggestion
        $hookNamesByAlias = static::getCanonicalHookNames();
        $loweredName = strtolower($hookName);
        if (isset($hookNamesByAlias[$loweredName])) {
            // return the canonical name (?)
            return $hookNamesByAlias[$loweredName];
```

PierreRambaud commented on classes/Hook.php:
```diff
@@ -995,4 +1004,237 @@ private static function getHookRegistry()
+        $customer = $context->customer;
+
+        $cache_id = self::MODULE_LIST_BY_HOOK_KEY
+            . ($shop instanceof Shop && isset($shop->id) ? '_' . $shop->id : '')
+            . ($customer instanceof Customer ? '_' . $customer->id : '');
```
```suggestion
            . (isset($shop->id) ? '_' . $shop->id : '')
            . (isset($customer->id) ? '_' . $customer->id : '');
```
Instanceof checks are useless because of isset

matks commented:
I have tested and validated the 4 scenarios provided by @eternoendless

matks commented:
Thank you @eternoendless

ProteanCode commented:
What is the correct way to make it working with both pre & post 1.7.7.0 instances?
